File: messages/ChatServerMessage.py
from .ChatAbstractServerMessage import ChatAbstractServerMessage
from CustomDataWrapper import CustomDataWrapper

class ChatServerMessage(ChatAbstractServerMessage):
    def __init__(self, channel = 0, content = '', timestamp = 0, fingerprint = '', senderId = 0, senderName = '', prefix = '', senderAccountId = 0):
        super().__init__(channel, content, timestamp, fingerprint)
        self.senderId = senderId
        self.senderName = senderName
        self.prefix = prefix
        self.senderAccountId = senderAccountId

    def serialize(self, output: CustomDataWrapper):
        super().serialize(output)
        if self.senderId < -9007199254740990 or self.senderId > 9007199254740990:
            raise ValueError("Forbidden value (" + str(self.senderId) + ") on element senderId.")
        output.writeDouble(self.senderId)
        output.writeUTF(self.senderName)
        output.writeUTF(self.prefix)
        if self.senderAccountId < 0:
            raise ValueError("Forbidden value (" + str(self.senderAccountId) + ") on element senderAccountId.")
        output.writeInt(self.senderAccountId)

    def deserialize(self, input):
        super().deserialize(input)
        self.senderId = input.readDouble()
        self.senderName = input.readUTF()
        self.prefix = input.readUTF()
        self.senderAccountId = input.readInt()

    def getMessageId(self) -> int:
        return 881
    
    def __str__(self) -> str:
        return f"channel: {self.channel}, content: {self.content}, timestamp: {self.timestamp}, fingerprint: {self.fingerprint}, senderId: {self.senderId}, senderName: {self.senderName}, prefix: {self.prefix}, senderAccountId: {self.senderAccountId}"
    

File: game/network/messages/ChatServerMessage.go
package game

import (
	"sniffsniff/utils"
)

type ChatServerMessage struct {
	BaseChatServerMessage ChatAbstractServerMessage
	SenderId              uint64
	SenderName            string
	Prefix                string
	SenderAccountId       uint32
}

func (message ChatServerMessage) GetId() int {
	return 881
}

func (message *ChatServerMessage) Deserialize(buffer *utils.Buffer) error {
	message.BaseChatServerMessage.Deserialize(buffer)
	senderId, err := buffer.ReadULong()
	if err != nil {
		return err
	}
	message.SenderId = senderId
	senderName, err := buffer.ReadUTF()
	if err != nil {
		return err
	}
	message.SenderName = senderName
	prefix, err := buffer.ReadUTF()
	if err != nil {
		return err
	}
	message.Prefix = prefix
	senderAccountId, err := buffer.ReadUInt()
	if err != nil {
		return err
	}
	message.SenderAccountId = senderAccountId
	return nil
}

func (message ChatServerMessage) String() string {
	return "ChatServerMessage{ChatAbstractServerMessage: " + message.BaseChatServerMessage.String() + ", SenderId: " + utils.ULongToString(message.SenderId) + ", SenderName: " + message.SenderName + ", Prefix: " + message.Prefix + ", SenderAccountId: " + utils.UIntToString(message.SenderAccountId) + "}"
}

func CreateChatServerMessage() FinalMessage {
	return &ChatServerMessage{
		BaseChatServerMessage: ChatAbstractServerMessage{
			ChannelId:   0,
			Content:     "",
			Timestamp:   0,
			Fingerprint: ""},
		SenderId:        0,
		SenderName:      "",
		Prefix:          "",
		SenderAccountId: 0,
	}
}
